lib/engine/game/g_18_uruguay/nationalization.rb

```diff
@@ -0,0 +1,286 @@
+          # train
+          corp.trains.each { |train| train.owner = corp_fce }
+          corp_fce.trains.concat(corp.trains)
+          corp.trains.clear
+          corp_fce.trains.each { |t| t.operated = false }
+        end
+
+        def swap_token(survivor, nonsurvivor, old_token)
+          city = old_token.city
+          exist = city.tokens.find { |token| token&.corporation == survivor }
+          nonsurvivor.tokens.delete(old_token)
+          if exist
+            @log << "Token removed in #{city.hex.id} since FCE already have one token in that location"
+            city.tokens[city.tokens.find_index(old_token)] = nil
+            return nil
+          end
+          new_token = survivor.next_token
+          @log << "Replaced #{nonsurvivor.name} token in #{city.hex.id} with #{survivor.name} token"
+          new_token.place(city)
+          city.tokens[city.tokens.find_index(old_token)] = new_token
+          new_token
+        end
+
+        def corps_to_nationalize
+          @round.entities.select { |entity| entity.loans.size.positive? && entity != @rptla }
+        end
+
+        def start_merge(originatior, _entity_a, _entity_b)
+          candidates = corps_to_nationalize
+          corp_a = nil
+          corp_b = nil
+          corps = []
+          if candidates.size.positive?
+            corp_a = candidates.shift
+            corps.append(corp_a)
+          end
+          if candidates.size.positive?
+            corp_b = candidates.shift
+            corps.append(corp_b)
+          end
+          @merge_data = {
+            holders: share_holder_list(originatior, corps),
+            corps: corps,
+            secondary_corps: [],
+            home_tokens: [],
+            tokens: [],
+            candidates: candidates,
+          }
+
+          if corp_a.nil?
+            @fce.close!
+            @corporations.delete(@fce)
+            return
+          end
+
+          @fce.ipoed = true
+          fce_share_price = compute_merger_share_price(corp_a, corp_b)
+          @fce.floatable = true
+          @stock_market.set_par(@fce, fce_share_price)
+          after_par(@fce)
+
+          acquire_shares_in_fce(@fce, @merge_data)
+        end
+
+        def nationalization_final_export!
+          @log << '  Nationalization: Final Export to be implemented'
+        end
+
+        def nationalization_close_rptla!
+          @log << '  Nationalization: RPTLA closes'
+          corporation = @rptla
+          corporation.share_holders.keys.each do |share_holder|
+            shares = share_holder.shares_of(corporation)
+            bundle = ShareBundle.new(shares)
+            sell_shares_and_change_price(bundle) unless corporation == share_holder
+          end
+          @rptla.close!
+          @corporations.delete(@rptla)
+        end
+
+        def close_rptla_private!
+          sym = 'JOHN'
+          company = @companies.find { |comp| comp.sym == sym }
```

Comment:
```suggestion
          company = company_by_id(sym)
```

`company_by_id` (and some other `_by_id` methods) is set up to cache `@companies`, storing a hash keyed on the sym, so that is a more efficient look up than linearly searching the array.